group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1190,10 +1190,11 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
             .setHeartbeatIntervalMs(consumerGroupHeartbeatIntervalMs);
 
         // The assignment is only provided in the following cases:
-        // 1. The member reported its owned partitions;
-        // 2. The member just joined or rejoined to group (epoch equals to zero);
-        // 3. The member's assignment has been updated.
-        if (ownedTopicPartitions != null || memberEpoch == 0 || hasAssignedPartitionsChanged(member, updatedMember)) {
+        // 1. The member sent a full request. It does so when joining or rejoining the group; or
+        //    on any errors (e.g. timeout).
+        // 2. The member's assignment has been updated.
+        boolean isFullRequest = memberEpoch == 0 || (rebalanceTimeoutMs != -1 && subscribedTopicNames != null && ownedTopicPartitions != null);
```

Comment:
nice, perfectly aligned with the client side. Just for the record, along with the `rebalanceTimeout`, `topics` and `assignment`, the client will also include the server assignor in any full request, but only if it's configured, so agree on not including it here.

Reply:
I think I'm missing how 
```(rebalanceTimeoutMs != -1 && subscribedTopicNames != null && ownedTopicPartitions != null)```
maps to 
```//  It does so when joining or rejoining the group; or on any errors (e.g. timeout).```

Are we instead saying that this particular case -- ie setting all these 3 fields indicates a "full" request? I think we should update the comment to make this a bit clearer.

Reply:
@jolshan Yes. You're understanding is correct. I updated the comment to explain it as it was explained in the description.